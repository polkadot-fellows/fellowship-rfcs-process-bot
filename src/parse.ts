const parse = (body: string) => {
  const match = body.match(/\/bot\s+(\w+)(.*)/)
  if (!match) {
    return {
      getArg: () => undefined
    }
  }

  const [, cmd, args] = match

  // use csv parser to handle quoted strings
  const argsArr: string[] = args.trim().split(/\s+/)

  const namedArgs: Record<string, string> = {}
  const unnamedArgs: string[] = []

  const normalizedNamedArgs: Record<string, string> = {}

  const normalize = (key: string) => {
    return key.trim().toLowerCase().replaceAll('-', '').replaceAll('_', '')
  }

  for (const arg of argsArr) {
    if (arg.trim().length === 0) {
      continue
    }
    if (arg.includes('=')) {
      const [key, value] = arg.split('=')
      const trimmedValue = value.trim()
      namedArgs[key.trim()] = trimmedValue
      normalizedNamedArgs[normalize(key)] = trimmedValue
    } else {
      unnamedArgs.push(arg)
    }
  }

  return {
    cmd,
    namedArgs,
    unnamedArgs,
    rawArgs: args,
    getArg(key: string) {
      return normalizedNamedArgs[normalize(key)]
    }
  }
}

export default parse
