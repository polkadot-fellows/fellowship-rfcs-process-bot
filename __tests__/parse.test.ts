import { expect, test } from '@jest/globals'
import parse from '../src/parse'

test('parse', () => {
  expect(parse('')).toMatchSnapshot()
  expect(parse('cmd')).toMatchSnapshot()
  expect(parse('/bot')).toMatchSnapshot()
  expect(parse('/bot cmd')).toMatchSnapshot()
  expect(parse('/bot cmd a b c')).toMatchSnapshot()
  expect(parse('/bot test val=123 val2=val2 val3')).toMatchSnapshot()

  const { getArg } = parse('/bot test some_value_1=1 SomeValue2=2 some-value-3=3 sOmev-al_uE4=4')
  expect(getArg('somevalue1')).toEqual('1')
  expect(getArg('some_value_2')).toEqual('2')
  expect(getArg('SomeValue3')).toEqual('3')
  expect(getArg('some-value-4')).toEqual('4')
})
