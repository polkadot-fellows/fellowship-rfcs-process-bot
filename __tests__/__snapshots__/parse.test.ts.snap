// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse 1`] = `
{
  "getArg": [Function],
}
`;

exports[`parse 2`] = `
{
  "getArg": [Function],
}
`;

exports[`parse 3`] = `
{
  "getArg": [Function],
}
`;

exports[`parse 4`] = `
{
  "cmd": "cmd",
  "getArg": [Function],
  "namedArgs": {},
  "rawArgs": "",
  "unnamedArgs": [],
}
`;

exports[`parse 5`] = `
{
  "cmd": "cmd",
  "getArg": [Function],
  "namedArgs": {},
  "rawArgs": " a b c",
  "unnamedArgs": [
    "a",
    "b",
    "c",
  ],
}
`;

exports[`parse 6`] = `
{
  "cmd": "test",
  "getArg": [Function],
  "namedArgs": {
    "val": "123",
    "val2": "val2",
  },
  "rawArgs": " val=123 val2=val2 val3",
  "unnamedArgs": [
    "val3",
  ],
}
`;
